File: kattis/iceland/fibonaccigjof.cpp
#pragma GCC optimize("Ofast","unroll-loops")
#pragma GCC target("avx2,fma")
#include <bits/stdc++.h>
using namespace std;
#define rep(i,a,b) for (__typeof(a) i=(a); i<(b); ++i)
#define iter(it,c) for (__typeof((c).begin()) \
  it = (c).begin(); it != (c).end(); ++it)
typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;
typedef long long ll;
typedef unsigned long long ull;
typedef vector<ull> vll;
typedef vector<vll> vvll;
typedef vector<vvll> vvvll;
#define pb push_back
#define fs first
#define sc second

const int INF = ~(1<<31);

const double EPS = 1e-9;
const double pi = acos(-1);
typedef vector<vi> vvi;
typedef vector<vii> vvii;
template <class T> T smod(T a, T b) {
  return (a % b + b) % b; }
#define mod 1000000007

vvvll mats(40, vvll(2, vll(2)));
vll fib;
ll c = 0;
//vll fib;

vvll matrixMult(vvll a, vvll b){
    vvll res(2, vll(2));
    rep(i, 0, 2){
        rep(j, 0, 2){
            rep(k, 0, 2){
                res[i][j] += (a[i][k] * b[k][j]) % mod;
            }
            res[i][j] %= mod;
        }
    }
    return res; //skoða
}

vll vecMult(vvll mat, vll vec){
    vll res(2);
    rep(i, 0, 2){
        rep(k, 0, 2){
            res[i] += (mat[i][k] * vec[k]) % mod;
        }
        res[i] %= mod;
    }

    //cout << res[0] << endl;

    return res; //skoða
}

vvll power(ull d){

    vvll res = mats[0];

    rep(i, 0, 39){
        if((d >> i) & 1){
            res = matrixMult(res, mats[i+1]);
        }
    }

    return res;
}

void genMats(){
    mats[0][0][0] = mats[0][1][1] = 1;
    mats[0][0][1] = mats[0][1][0] = 0;
    mats[1][0][1] = mats[1][1][1] = mats[1][1][0] = 1;
    mats[1][0][0] = 0;
    for(int e=2; e < 40; e++){
        mats[e] = matrixMult(mats[e-1], mats[e-1]);
    }
}

class segmentTree{
    public:
        ull n;
        vll lo, hi, delta;
        vvll sum;

    segmentTree(ull n){
        this-> n = n;
        lo.resize(4*n+1);
        hi.resize(4*n+1);
        delta.resize(4*n+1);
        sum.resize(4*n+1, vll(2));
        init(1, 0, n-1);
    }

    void init(ull i, ull a, ull b){
        lo[i] = a;
        hi[i] = b;

        if (a == b){
            sum[i][0] = vecMult(power(fib[c]), {0, 1})[0];
            sum[i][1] = vecMult(power(fib[c]), {0, 1})[1];
            //cout << sum[i][0] << ' ' << sum[i][1] << endl;
            c++;
            return;
        }

        ll m = (a+b-1)/2;
        init(2*i, a, m);
        init(2*i+1, m+1, b);

        update(i);
    }

    void prop(ull i){
        //cout << delta[i] << endl;
        delta[2*i] += delta[i];
        delta[2*i+1] += delta[i];
        delta[i] = 0;
        //update(i);
    }

    void update(ull i){
        vll leftSum = vecMult(power(delta[2*i]), sum[2*i]);
        vll rightSum = vecMult(power(delta[2*i+1]), sum[2*i+1]); 

        sum[i][0] = (leftSum[0] + rightSum[0]) % mod;
        sum[i][1] = (leftSum[1] + rightSum[1]) % mod;
    }

    void add(ull i, ull a, ull b, ull val){
        if (b < lo[i] || hi[i] < a)
            return;
        if (a <= lo[i] && hi[i] <= b){
            //cout << i << endl;
            delta[i] += val;
            return;
        }

        prop(i);

        add(2*i, a, b, val);
        add(2*i+1, a, b, val);

        update(i);
    }

    ull sumQuary(ull i, ull a, ull b){
        //cout << i << endl;
        if (b < lo[i] || hi[i] < a)
            return 0;
        if (a <= lo[i] && hi[i] <= b){
            return vecMult(power(delta[i]), sum[i])[0] % mod;
        }

        prop(i);

        ull sum1 = sumQuary(2*i, a, b) % mod;
        ull sum2 = sumQuary(2*i+1, a, b) % mod;

        update(i);

        return (sum1 + sum2) % mod;
    }

};

int main(){
    ull n, m, t, l, r, d;
    cin >> n >> m;
    genMats();
    fib.resize(n);
    rep(i, 0, n) cin >> fib[i];
    segmentTree st(n);
    rep(i, 0, m){
        cin >> t >> l >> r;
        l -= 1, r -= 1;
        if (t == 1){
            cin >> d;
            st.add(1, l, r, d);
        }
        else{
            cout << st.sumQuary(1, l, r) % mod << '\n';
        }
    }
    return 0;
}

File: kattis/iceland/fibonaccigjof.py
import sys

mod = 10**9 + 7

mats = [[[0 for i in range(2)] for i in range(2)] for _ in range(35)]

fib = []
c = [0]

def matrixMult(a, b):
    res = [[0, 0], [0, 0]]
    for i in range(2):
        for j in range(2):
            for k in range(2):
                res[i][j] += (a[i][k] * b[k][j]) % mod
            res[i][j] %= mod
    return res

def vecMult(mat, vec):
    res = [0, 0]
    for i in range(2):
        for k in range(2):
            res[i] += (mat[i][k] * vec[k]) % mod
        res[i] %= mod
    
    #print res[0]

    return res

def power(d):
    res = mats[0]

    #for i in bin(d)[2:]: print i

    for i in range(34):
        if (d>>i) & 1:
            res = matrixMult(res, mats[i+1])
    
    return res

def genMats():
    mats[0] = [[1, 0], [0, 1]]
    mats[1] = [[0, 1], [1, 1]]

    for e in range(2, 35):
        mats[e] = matrixMult(mats[e-1], mats[e-1])

class segmentTree:
    def __init__(self, n):
        self.n = n
        self.lo = [0 for i in range(4*n+1)]
        self.hi = [0 for i in range(4*n+1)]
        self.delta = [0 for i in range(4*n+1)]
        self.sum = [[0, 0] for i in range(4*n+1)]
        self.init(1, 0, n-1)

    def init(self, i, a, b):
        self.lo[i] = a
        self.hi[i] = b

        if a == b:
            self.sum[i][0] = vecMult(power(fib[c[0]]), [0, 1])[0]
            self.sum[i][1] = vecMult(power(fib[c[0]]), [0, 1])[1]
            #print self.sum[i][0], self.sum[i][1]
            c[0] += 1
            return
        
        m = (a+b-1)//2

        self.init(2*i, a, m)
        self.init(2*i+1, m+1, b)

        self.update(i)
    
    def prop(self, i):
        #print self.delta[i]
        self.delta[2*i] += self.delta[i]
        self.delta[2*i+1] += self.delta[i]
        self.delta[i] = 0

    def update(self, i):
        leftSum = vecMult(power(self.delta[2*i]), self.sum[2*i])
        rightSum = vecMult(power(self.delta[2*i+1]), self.sum[2*i+1])

        self.sum[i][0] = (leftSum[0] + rightSum[0]) % mod
        self.sum[i][1] = (leftSum[1] + rightSum[1]) % mod

    def add(self, i, a, b, val):
        if b < self.lo[i] or self.hi[i] < a:
            return
        if a <= self.lo[i] and self.hi[i] <= b:
            self.delta[i] += val
            return
        
        self.prop(i)

        self.add(2*i, a, b, val)
        self.add(2*i+1, a, b, val)

        self.update(i)
    
    def sumQuary(self, i, a, b):
        #print i
        if b < self.lo[i] or self.hi[i] < a:
            return 0
        if a <= self.lo[i] and self.hi[i] <= b:
            return vecMult(power(self.delta[i]), self.sum[i])[0] % mod
        
        self.prop(i)

        sum1 = self.sumQuary(2*i, a, b) % mod
        sum2 = self.sumQuary(2*i+1, a, b) % mod

        self.update(i)

        return (sum1 + sum2) % mod

n, m, t, l, r, d = 0, 0, 0, 0, 0, 0

n, m = map(int, sys.stdin.readline().split())

genMats()

fib = list(map(int, sys.stdin.readline().split()))

st = segmentTree(n)

for i in range(m):
    a = list(map(int, sys.stdin.readline().split()))
    if len(a) == 4:
        t, l, r, d = map(int, a)
        l -= 1
        r -= 1
        st.add(1, l, r, d)
    else:
        t, l, r = map(int, a)
        l -= 1
        r -= 1
        print st.sumQuary(1, l, r) % mod